### PR TITLE
Fix Circle CI Integration | Circle CI API v1.1

### LINKED ISSUE #791: Circle CI support partly broken
Hello and thank you for your great work. :)

The [shields.io](https://shields.io) website claims to support private projects on [Circle CI](https://circleci.com) via Circle's support for [status tokens](https://circleci.com/docs/status-badges/#access-control), which only have access to the build status and nothing else.

This is super awesome. It just doesn't seem to work. All I can see is "build | invalid".

To demonstrate, first the pattern from the Shields website, then a non-working and working example.

Shields pattern:
https://img.shields.io/circleci/token/YOURTOKEN/project/BrightFlair/PHP.Gt/master.svg

Non-working sample using Shields:
https://img.shields.io/circleci/token/acc2259949dc35d324ffff96186e5874a4b1ef4f/project/sitecues/sitecues-core/master.svg

Working sample using Circle:
https://circleci.com/gh/sitecues/sitecues-core.svg?style=svg&circle-token=acc2259949dc35d324ffff96186e5874a4b1ef4f

### PR DESCRIPTION
**Pull request by RedSparr0w**
Fixes #791 @sholladay

url structure:
`/circleci/token/YOURTOKEN/project/github/RedSparr0w/node-csgo-parser/master.svg`

part of url | required | explanation
--- | --- | ---
/token/YOURTOKEN | no | token if private Repo
/github | yes | can either be github or bitbucket depending where repo hosted
/RedSparr0w | yes | Username
/node-csgo-parser | yes | Repository
/master | no | Branch

### CODE REVIEW

**Comment by levithomason** [comment]
@espadrine any chance we can get this merged?  Much appreciated.

File: try.html [hunk 1]
```diff
@@ -117,16 +117,16 @@ <h3 id="build"> Build </h3>
     <td><code>https://img.shields.io/magnumci/ci/96ffb83fa700f069024921b0702e76ff/new-meta.svg</code></td>
   </tr>
   <tr><th> CircleCI: </th>
-      <td><img src='/circleci/project/BrightFlair/PHP.Gt.svg' alt=''/></td>
-      <td><code>https://img.shields.io/circleci/project/BrightFlair/PHP.Gt.svg</code></td>
+      <td><img src='/circleci/project/github/RedSparr0w/node-csgo-parser.svg?maxAge=2592000' alt=''/></td>
```
Comment (espadrine) [inline try.html]: Could you remove maxAge from try.html please?

index.html is meant to be generated from try.html through `make website`, which I can do for you.

File: server.js [hunk 2]
```diff
@@ -4479,15 +4479,16 @@ cache(function(data, match, sendBadge, request) {
...
-  var format = match[4];
+  var token = match[1];
+  var type = match[2]; // github OR bitbucket
+  var userRepo = match[3];  // eg, `RedSparr0w/node-csgo-parser`.
+  var branch = match[4]
+  var format = match[5];
 
   // Base API URL
-  var apiUrl = 'https://circleci.com/api/v1/project/' + userRepo;
+  var apiUrl = 'https://circleci.com/api/v1.1/project/'+ type +"/"+ userRepo;
```
Comment (espadrine) [inline server.js]: Could you put spaces around the `+` operators please?

File: server.js [hunk 3]
```diff
@@ -4479,15 +4479,16 @@ cache(function(data, match, sendBadge, request) {
...
 // https://circleci.com/api/v1/project/BrightFlair/PHP.Gt?circle-token=0a5143728784b263d9f0238b8d595522689b3af2&limit=1&filter=completed
-camp.route(/^\/circleci\/(?:token\/(\w+))?[+\/]?project\/([^\/]+\/[^\/]+)(?:\/(.*))?\.(svg|png|gif|jpg|json)$/,
+camp.route(/^\/circleci\/(?:token\/(\w+))?[+\/]?project\/(github|bitbucket)\/([^\/]+\/[^\/]+)(?:\/(.*))?\.(svg|png|gif|jpg|json)$/,
 cache(function(data, match, sendBadge, request) {
-  var token = match[1]
-  var userRepo = match[2];  // eg, `BrightFlair/PHP.Gt`.
-  var branch = match[3];
-  var format = match[4];
+  var token = match[1];
+  var type = match[2]; // github OR bitbucket
```
Comment (espadrine) [inline server.js]: The addition of this piece of the URL will break existing links. Could you make that part optional, defaulting to GitHub, please?

**Comment by RedSparr0w** [comment]
Updated with requested changes.

**Comment by espadrine** [comment]
Thanks a lot!

**Comment by sholladay** [comment]
I haven't been able to reproduce a working example using a [build status token](https://circleci.com/docs/status-badges/#access-control) with the img.shields.io server that's currently online. The home page seems to indicate this has been deployed. @espadrine is this correct?

A full API token does work, but using those is dangerous for this use case.

Here is an example that ought to say "passing", but instead says "invalid".
https://img.shields.io/circleci/token/acc2259949dc35d324ffff96186e5874a4b1ef4f/project/github/sitecues/sitecues-core.svg

Here is the equivalent directly from Circle, which works.
https://circleci.com/gh/sitecues/sitecues-core.svg?style=svg&circle-token=acc2259949dc35d324ffff96186e5874a4b1ef4f

**Comment by levithomason** [comment]
FWIW, I had to add a token with `All` permissions.  Neither the `status` nor the `build artifact` token scopes worked.

**Comment by RedSparr0w** [comment]
`https://circleci.com/api/v1.1/project/github/sitecues/sitecues-core?circle-token=acc2259949dc35d324ffff96186e5874a4b1ef4f&limit=1&filter=completed` just gives:
```json
{
  "message" : "Permission denied"
}
```

but seems to work using view-builds scope - gives permission denied error for `status` token or project not found for no token/invalid token
Image:  | Api: ["view-builds" token](https://circleci.com/api/v1.1/project/github/RedSparr0w/Pokemon-GO-node-api?circle-token=8b8a7fc4ea7b70b1a4f7e698a7833e68a8ed54c8&limit=1&filter=completed)
Image:  | Api: ["status" token](https://circleci.com/api/v1.1/project/github/RedSparr0w/Pokemon-GO-node-api?circle-token=646538d3aff72301bba42d7c9b9c2da074f6ebaf&limit=1&filter=completed)
Image:  | Api: ["all" token](https://circleci.com/api/v1.1/project/github/RedSparr0w/Pokemon-GO-node-api?circle-token=95b25ddddac66a810e18c6244df53fc448c032d7&limit=1&filter=completed)
Image:  | Api: [no token](https://circleci.com/api/v1.1/project/github/RedSparr0w/Pokemon-GO-node-api?limit=1&filter=completed)

**Comment by levithomason** [comment]
Confirmed creating a token with `Build Artifacts` works, thanks.  Not sure why it didn't work for me previously.

**Comment by RedSparr0w** [comment]
Do you think it would be better to show the message _(if available)_ when invalid?
Just so you have a idea of what the problem could be.

example:


-vs-

**Comment by levithomason** [comment]
Oh man, that would be immensely helpful

**Comment by levithomason** [comment]
Perhaps even a query param to view the raw JSON response. That way, it would be standardised for all badge providers. Something like `?json` or `? debug`.

**Comment by RedSparr0w** [comment]
I've implemented what i was saying above in #816

But to dump the json for all badges would probably require quite a massive amount of changes.

**Comment by espadrine** [comment]
> Perhaps even a query param to view the raw JSON response. That way, it would be standardized for all badge providers.

A few vendor data is obtained through non-JSON formats, some over several requests.

**Pull request merged**